src/zap/zapinfo.cpp

```diff
@@ -2083,6 +2102,59 @@ void ZapInfo::GetProfilingHandle(BOOL                      *pbHookFunction,
     *pbIndirectedHandles = TRUE;
 }
 
+DWORD FilterMethodAttribsForIsSupported(DWORD attribs, CORINFO_METHOD_HANDLE ftn, ICorDynamicInfo* pJitInfo)
+{
+    if (attribs & CORINFO_FLG_JIT_INTRINSIC)
+    {
+        // Do not report the get_IsSupported method as an intrinsic. This will turn the call into a regular
+        // call. We also make sure none of the hardware intrinsic method bodies get pregenerated in crossgen
+        // (see ZapInfo::CompileMethod) but get JITted instead. The JITted method will have the correct
+        // answer for the CPU the code is running on.
+
+        const char* namespaceName;
+        const char* className;
+        const char* enclosingClassName;
+        const char* methodName = pJitInfo->getMethodNameFromMetadata(ftn, &className, &namespaceName, &enclosingClassName);
+
+        // If it's not the IsSupported method, we're done
+        if (strcmp(methodName, "get_IsSupported") != 0)
+            return attribs;
+
+        bool isX86intrinsic = strcmp(namespaceName, "System.Runtime.Intrinsics.X86") == 0;
+
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+        // If it's IsSupported on Sse/Sse2, we can expand unconditionally since this is reliably
+        // available everywhere.
+        if (isX86intrinsic
+            && (
+                strcmp(className, "Sse") == 0 || strcmp(className, "Sse2") == 0
+                || (
+                    strcmp(className, "X64") == 0
+                    && (
+                        strcmp(enclosingClassName, "Sse") == 0 || strcmp(enclosingClassName, "Sse2") == 0
+                        )
+                    )
+                )
+            )
+            return attribs;
+#endif // defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+
+        // If it's another get_IsSupported method on some other ISA class, do not report as intrinsic
+        if (isX86intrinsic
+            || strcmp(namespaceName, "System.Runtime.Intrinsics.Arm.Arm64") == 0
+            || strcmp(namespaceName, "System.Runtime.Intrinsics") == 0)
+        {
+            attribs &= ~CORINFO_FLG_JIT_INTRINSIC;
+
+            // The intrinsic just recursively calls itself in IL so it wouldn't inline anyway,
```

Comment:
I think this comment should be updated. The importer does recognize and handle recursive calls and inlines the intrinsic into itself. However, my understanding is it is the above checks which avoids AOT'ing the intrinsic bodies that allows this to "work".

Reply:
If the importer were to inline the recursive call, RyuJIT would have to generate an infinite loop or a `throw StackOverflowException`. I didn't see that happening, which is why I assume it doesn't currently inline.

Reply:
The intrinsic path has special logic in the importer and tracks recursive named intrinsics as `mustExpand`.

We then use that `mustExpand` variable to determine if we should return `nullptr` or intrinsic node.

For most of the intrinsics, this only impacts whether the intrinsic is "inlined" or a call (with it's own body "inlined" into itself), but for a few select intrinsics, they will emit a jump table instead (intrinsics that require a constant parameter).

This is how reflection/etc is supported as well.

Reply:
Right; I think the comment could be reworded as "Intrinsics recursively invoke themselves (to support reflection, debugging etc.), so inlining would be pointless if we are not going to recognize the intrinsic."